File: AStar.py
__author__ = 'Douglas Lewis @ douglasdlewis2012@gmail.com'

from collections import defaultdict
from heapq import heappop, heappush
from timeit import timeit
import random
import cProfile as profile
'''
Given two five letter words, A and B, and a dictionary of five letter words,
find a shortest transformation from A to B, such that only one letter can be
changed at a time and all intermediate words in the transformation must exist in the dictionary.
 
For example, if A and B are "smart" and "brain", the result may be:
smart
    start
    stark
    stack
    slack
    black
    blank
    bland
    brand
    braid
brain
 
Your implementation should take advantage of multiple CPU cores.
Please also include test cases against your algorithm.
Your solution should produce the words that make up the shortest path itself, not just the count.
'''

class AStar:

    def __init__(self):
        print('initializing')

    class Node:

        def __init__(self, f, g, word, previous):
            self.f = f
            self.g = g
            self.word = word
            self.previous = previous

        def __lt__(self, other):
            return self.f < other.f

    def heuristic(self,word, end):
        return sum([a != b for a,b in zip(word,end)])

    def generate_neighbors(self, words):
        holder = object()
        match = defaultdict(list)
        neighbors = defaultdict(list)
        for w in words:
            for i in range(len(w)):
                p = tuple(holder if i == j else c for j, c in enumerate(w))
                m = match[p]
                m.append(w)
                neighbors[w].append(m)
        return neighbors

    def reconstruct_path(self, node):
        path = [node.word]
        while node.previous != None:
            temp = node.previous
            node = node.previous
            path.append(temp.word)
        p = path[::-1]

        return ' '.join(p)


    def a_star_search(self, start, end, words):
        neighbors = self.generate_neighbors(words)
        openSet = set([start])
        closedSet = set()
        heap = [AStar.Node(self.heuristic(start, end),0,start, None)]

        while len(heap) > 0:
            node = heappop(heap)
            if node.word == end:
                return self.reconstruct_path(node)

            openSet.remove(node.word)
            closedSet.add(node.word)

            g = node.g + 1
            for n in neighbors[node.word]:
                for w in n:
                    if w in closedSet:
                        continue

                    if w not in openSet:
                        new_node = AStar.Node(self.heuristic(w, end) + g, g, w, node)
                        heappush(heap, new_node)
                        openSet.add(w)

        msg = 'No way found! between %s - %s' %(start, end)
        return msg

    def get_words(self, size=5):
        all_words = [w.strip().lower() for w in open('words.txt')]

        print('total words: ' , len(all_words))
        words_each_size = {len(w):0 for w in all_words}

        for w in all_words:
            words_each_size[len(w)] += 1

        print(words_each_size)
        sized_words = [w for w in all_words if len(w) == size]

        return sized_words

    def time_a_star(self, start, end, words):
        print('start: %s end: %s ladder: %s' %(start, end, self.a_star_search(start, end, words)))

    def test_time_searches(self):
        words = self.get_words()
        start = words[random.randint(1,len(words))]
        end =  words[random.randint(1,len(words))]
        print(timeit(lambda:print(self.time_a_star(start, end, words)), number=1))


    def runTests(self):
        profile.runctx('self.test_time_searches()', globals(), locals())

    def test(self, start, end):
        self.start = start
        self.end = end
        words = self.get_words()
        print(self.a_star_search(start, end, words))

File: AStar.java
package com.douglaslewis;

import java.io.File;
import java.util.*;


public class AStar {
    private final String HOLDER = "_";

    private class Node implements Comparable<Node> {
        private int f;
        private int g;
        private String word;
        private Node previous;

        public Node(int f, int g, String word, Node previous) {
            this.f = f;
            this.g = g;
            this.word = word;
            this.previous = previous;
        }

        public void printNode() {
            StringBuilder builder = new StringBuilder();
            builder.append("f: ");
            builder.append(this.f);
            builder.append(" g: ");
            builder.append(this.g);
            builder.append(" word: ");
            builder.append(this.word);
            builder.append(" previous: ");
            builder.append(this.previous);
            System.out.println(builder.toString());
        }

        public int getF() {
            return f;
        }

        public int getG() {
            return g;
        }

        public String getWord() {
            return word;
        }

        public Node getPrevious() {
            return previous;
        }

        @Override
        public int compareTo(Node o) {
            String otherWord = ((Node) o).word;
            return this.word.compareTo(otherWord);
        }


    }


    public AStar() {
    }

    private int heuristic(String word, String end) {

        int sum = 0;
        char[] wordsArray = word.toCharArray();
        char[] endArray = end.toCharArray();

        for (int i = 0; i < word.length(); i++) {
            if (wordsArray[i] != endArray[i])
                sum++;
        }
        return sum;
    }


    private List<String> getWords(int size) {
        List<String> allWords = new ArrayList<String>();
        List<String> sizedWords = new ArrayList<String>();

        try {
            Scanner s = new Scanner(new File("words.txt"));

            while (s.hasNext()) {
                allWords.add(s.next().trim());
            }

            s.close();
        } catch (Exception e) {
            e.printStackTrace();
        }



        for (String word : allWords) {
            if (word.length() == size) {
                    sizedWords.add(word.toLowerCase());
            }
        }

        return sizedWords;
    }


    private Map<String, List<List<String>>> generateNeighbors(List<String> words) {
        Map<String, List<List<String>>> neighborhoods = new HashMap<>();
        Map<String, List<String>> match = new HashMap<>();

        for (String w : words) {
            List<List<String>> neihborhood = new ArrayList<>();
            for (int i = 0; i < w.length(); i++) {

                String p = getPattern(w,i);

                List<String> m = match.get(p);

                if (m == null) {
                    match.put(p, new ArrayList<>());
                    m = match.get(p);
                }

                m.add(w);
                neihborhood.add(m);
                neighborhoods.put(w, neihborhood);

            }
        }

        return neighborhoods;
    }

    private String getPattern(String w, int i ){
        String p = new String();

        for (int j = 0; j < w.length(); j++) {
            if (i == j) {
                p += HOLDER;
            } else {
                p += w.substring(j, j + 1);
            }
        }

        return p;

    }

    private String reconstructPath(Node node) {
        List<String> path = new ArrayList<String>();

        path.add(node.getWord());
        while (node.previous != null) {
            Node temp = node.previous;
            node = node.previous;
            path.add(temp.word);
        }

        Collections.reverse(path);


        StringBuilder pathString = new StringBuilder();
        for (String n : path) {
            pathString.append(n + " ");

        }

        return pathString.toString();
    }

    public String aStarSearch(String start, String end, List<String> words) {
        Map<String, List<List<String>>> neighbors = generateNeighbors(words);
        Set<String> openSet = new HashSet<>();
        openSet.add(start);

        Set<String> closedSet = new HashSet<>();
        PriorityQueue<Node> heap = new PriorityQueue<>();
        heap.add(new Node(heuristic(start, end), 0, start, null));


        while (!heap.isEmpty()) {
            Node node = heap.remove();
            if (node.getWord().equals(end)) {
                return reconstructPath(node);
            }

            openSet.remove(node.getWord());
            closedSet.add(node.getWord());

            int g = node.getG() + 1;
            List<List<String>> neighborhood = neighbors.get(node.getWord());
            for (List<String> localNeighbors : neighborhood) {
                for (String word : localNeighbors) {
                    if ((!closedSet.contains(word)) && (!openSet.contains(word))) {
                        Node newNode = new Node(heuristic(word, end) + g, g, word, node);
                        heap.add(newNode);
                        openSet.add(word);
                    }

                }

            }

        }

        return "No way found! between " + start + " - " + end;
    }

    public void runTests() {

    }

    public void test(String start, String end) {


        int size = 5;
        long startTime = System.nanoTime();
        List<String> words = getWords(size);
        System.out.println(aStarSearch(start, end, words));
        long endTime = System.nanoTime();
        System.out.println("Total time: " + (endTime - startTime) / 1.0e9);
        System.out.println("========================================");

    }
}
